rllib/core/rl_module/torch/tests/test_torch_marl_module.py

```diff
@@ -0,0 +1,275 @@
+            set(state[DEFAULT_POLICY_ID].keys()),
+            set(module[DEFAULT_POLICY_ID].get_state().keys()),
+        )
+
+        module2 = SimplePPOModule(config).as_multi_agent()
+        state2 = module2.get_state()
+        self.assertRaises(AssertionError, lambda: check(state, state2))
+
+        module2.set_state(state)
+        state2_after = module2.get_state()
+        check(state, state2_after)
+
+    def test_add_remove_modules(self):
+        # TODO (Avnish): Modify this test to make sure that the distributed
+        # functionality won't break the add / remove.
+
+        env_class = make_multi_agent("CartPole-v0")
+        env = env_class({"num_agents": 2})
+        config = get_shared_encoder_config(env)
+        module = SimplePPOModule(config).as_multi_agent()
+
+        module.add_module("test", SimplePPOModule(config))
+        self.assertEqual(set(module.keys()), {DEFAULT_POLICY_ID, "test"})
+        module.remove_module("test")
+        self.assertEqual(set(module.keys()), {DEFAULT_POLICY_ID})
+
+        # test if add works with a conflicting name
+        self.assertRaises(
+            ValueError,
+            lambda: module.add_module(DEFAULT_POLICY_ID, SimplePPOModule(config)),
+        )
+
+        module.add_module(DEFAULT_POLICY_ID, SimplePPOModule(config), override=True)
+
+    def test_rollouts(self):
+        for env_name in ["CartPole-v0", "Pendulum-v1"]:
+            for fwd_fn in ["forward_exploration", "forward_inference"]:
+                env_class = make_multi_agent(env_name)
+                env = env_class({"num_agents": 2})
+                config = get_shared_encoder_config(env)
+                module = SimplePPOModule(config).as_multi_agent()
+
+                def policy_map(agent_id):
+                    return DEFAULT_POLICY_ID
+
+                obs = env.reset()
+
+                tstep = 0
+                while tstep < 10:
+                    # go from agent_id to module_id, compute the actions in batches
+                    # and come back to the original per agent format.
+                    agent_obs = tree.map_structure(lambda x: {"obs": x}, obs)
+                    fwd_in = get_policy_data_from_agent_data(agent_obs, policy_map)
+                    fwd_in = tree.map_structure(
+                        lambda x: to_tensor(x) if isinstance(x, np.ndarray) else x,
+                        fwd_in,
+                    )
+
+                    if fwd_fn == "forward_exploration":
+                        fwd_out = module.forward_exploration(fwd_in)
+
+                        action = get_action_from_ma_fwd_pass(
+                            agent_obs, fwd_out, fwd_in, policy_map
+                        )
+
+                    elif fwd_fn == "forward_inference":
+                        # check if I sample twice, I get the same action
+                        fwd_out = module.forward_inference(fwd_in)
+
+                        action = get_action_from_ma_fwd_pass(
```

Comment:
Can we do this a bunch of times (not just twice?) Cartpole only has two actions, so this could fail some times, even if not deterministic.

Reply:
Oh, sorry, you do this already 10 times. I guess it's ok, then.